### PR TITLE
[RFR] Ssr

### PR DESCRIPTION
**Pull request by ThieryMichel**

### CODE REVIEW

File: src/api/controller/front/index.js [hunk 1]
```diff
@@ -1,33 +1,164 @@
...
+            userAgent: headers['user-agent'],
+        },
+    );
+    const history = createMemoryHistory(url);
+
+    ctx.body = await new Promise((resolve, reject) => {
+        match(
+            { history, routes, location: url },
+            (error, redirect, renderProps) => {
+                const store = configureStoreServer(
```
Comment (fzaninotto) [inline src/api/controller/front/index.js]: Can't you export that block to make it testable, and reduce the size of the current file ?

File: src/api/controller/index.js [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import path from 'path';
```
Comment (fzaninotto) [inline src/api/controller/index.js]: not used?

File: src/api/models/publishedDataset.js [hunk 3]
```diff
@@ -231,7 +231,8 @@ export default async (db) => {
 
     collection.findDistinctValuesForField = field => collection.distinct(`versions.${field}`);
 
-    collection.countByFacet = (field, value) => collection.count({ [`versions.${field}`]: value });
+    collection.countByFacet = (field, value) =>
```
Comment (fzaninotto) [inline src/api/models/publishedDataset.js]: async?

File: src/app/js/admin/App.js [hunk 4]
```diff
@@ -1,17 +1,18 @@
...
 import AppBar from './Appbar';
+import { Helmet } from 'react-helmet';
 
 export const AppComponent = ({ children }) => (
-    <MuiThemeProvider>
-        <div>
-            <AppBar />
+    <div>
+        <Helmet>
+            <title>{/https?:\/\/([\w-]+)/.exec(process.env.EZMASTER_PUBLIC_URL)[1]}</title>
```
Comment (fzaninotto) [inline src/app/js/admin/App.js]: please extract to standalone function

File: src/app/js/admin/clear/sagas.js [hunk 5]
```diff
@@ -42,7 +42,4 @@ export function* watchClearPublishedRequest() {
     yield takeLatest(CLEAR_PUBLISHED, handleClearPublishedRequest);
 }
 
-export default function* () {
-    yield fork(watchClearDatasetRequest);
-    yield fork(watchClearPublishedRequest);
-}
+export default [watchClearDatasetRequest, watchClearPublishedRequest];
```
Comment (fzaninotto) [inline src/app/js/admin/clear/sagas.js]: please revert

File: src/app/js/admin/preview/sagas.js [hunk 6]
```diff
@@ -3,7 +3,7 @@ import { fork } from 'redux-saga/effects';
 import publicationPreview from './publication/sagas';
 import fieldPreview from './field/sagas';
 
-export default function* () {
```
Comment (fzaninotto) [inline src/app/js/admin/preview/sagas.js:6]: same, please revert if not necessary

File: src/app/js/admin/removedResources/sagas/index.js [hunk 7]
```diff
@@ -2,7 +2,7 @@ import { fork } from 'redux-saga/effects';
 import loadRemovedResource from './loadRemovedResource';
 import restoreResource from './restoreResource';
 
-export default function* () {
-    yield fork(loadRemovedResource);
-    yield fork(restoreResource);
-}
+export default [
```
Comment (fzaninotto) [inline src/app/js/admin/removedResources/sagas/index.js]: same

File: src/app/js/admin/sagas.js [hunk 8]
```diff
@@ -15,19 +15,19 @@ import fetchSaga from '../fetch/sagas';
 import i18nSagas from '../i18n/sagas';
 import clearSaga from './clear/sagas';
 
-export default function* () {
+export default function*() {
     yield fork(exportSaga);
     yield fork(fetchSaga);
-    yield fork(fieldsSaga);
+    yield all(fieldsSaga.map(fork));
```
Comment (fzaninotto) [inline src/app/js/admin/sagas.js]: please revert here and down in this file

File: src/app/js/admin/upload/sagas/index.js [hunk 9]
```diff
@@ -3,7 +3,7 @@ import { fork } from 'redux-saga/effects';
 import uploadFile from './uploadFile';
 import uploadUrl from './uploadUrl';
 
-export default function* () {
-    yield fork(uploadFile);
-    yield fork(uploadUrl);
-}
+export default [
```
Comment (fzaninotto) [inline src/app/js/admin/upload/sagas/index.js]: revert

File: src/app/js/fields/sagas/index.js [hunk 10]
```diff
@@ -8,13 +8,13 @@ import changePosition from './changePosition';
...
-    yield fork(loadFields);
-    yield fork(removeField);
-    yield fork(saveField);
-    yield fork(validation);
-    yield fork(changeOperation);
-    yield fork(changePosition);
-    yield fork(loadPublication);
-    yield fork(configureField);
-}
+export default [
```
Comment (fzaninotto) [inline src/app/js/fields/sagas/index.js]: please revert

File: src/app/js/i18n/sagas/index.js [hunk 11]
```diff
@@ -3,7 +3,7 @@ import { fork } from 'redux-saga/effects';
 import setLanguage from './setLanguage';
 import initializeLanguage from './initializeLanguage';
 
-export default function* () {
-    yield fork(initializeLanguage);
-    yield fork(setLanguage);
-}
+export default [
```
Comment (fzaninotto) [inline src/app/js/i18n/sagas/index.js]: please revert

File: src/app/js/lib/queryStringToLiteral.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+export default (queryString = '') =>
```
Comment (fzaninotto) [inline src/app/js/lib/queryStringToLiteral.js]: you should use a module like `qs` for that.

File: src/app/js/public/App.js [hunk 13]
```diff
@@ -1,22 +1,13 @@
...
 
 export const AppComponent = ({ children }) => (
-    <MuiThemeProvider>
-        <div className="body">
-            {children}
-        </div>
-    </MuiThemeProvider>
+    <div className="body">
+        <Helmet>
+            <title>{/https?:\/\/([\w-]+)/.exec(process.env.EZMASTER_PUBLIC_URL)[1]}</title>
```
Comment (fzaninotto) [inline src/app/js/public/App.js]: please move to external function

File: src/app/js/public/Home.js [hunk 14]
```diff
@@ -99,6 +100,9 @@ export class HomeComponent extends Component {
         if (hasPublishedDataset) {
             return (
                 <div>
+                    <Helmet>
+                        <title>{/https?:\/\/([\w-]+)/.exec(process.env.EZMASTER_PUBLIC_URL)[1]}</title>
```
Comment (fzaninotto) [inline src/app/js/public/Home.js]: please move to external function

File: src/app/js/public/characteristic/sagas/index.js [hunk 15]
```diff
@@ -3,7 +3,7 @@ import { fork } from 'redux-saga/effects';
 import updateCharacteristic from './updateCharacteristic';
 import addCharacteristic from './addCharacteristic';
 
-export default function* () {
-    yield fork(updateCharacteristic);
-    yield fork(addCharacteristic);
-}
+export default [
```
Comment (fzaninotto) [inline src/app/js/public/characteristic/sagas/index.js]: please revert

File: src/app/js/public/overview/Overview.js [hunk 16]
```diff
@@ -39,58 +40,87 @@ const styles = {
...
+                                }
+                                title={
+                                    <a
+                                        href={getResourceUri(data)}
+                                        title={
+                                            columns.filter(
+                                                e => e.overview === 1,
+                                            ).length
+                                                ? data[
+                                                      columns.filter(
```
Comment (fzaninotto) [inline src/app/js/public/overview/Overview.js]: please refactor to avoid repetition

File: src/app/js/public/resource/Detail.js [hunk 17]
```diff
@@ -140,6 +140,9 @@ export const DetailComponent = ({
 
     return (
         <div className="detail">
+            <Helmet>
+                <title>{/https?:\/\/([\w-]+)/.exec(process.env.EZMASTER_PUBLIC_URL)[1]} - {sharingTitle || resource.uri}</title>
```
Comment (fzaninotto) [inline src/app/js/public/resource/Detail.js]: please extract

File: src/app/js/public/resource/sagas/index.js [hunk 18]
```diff
@@ -7,11 +7,11 @@ import addFieldToResource from './addFieldToResource';
...
 
-export default function* () {
-    yield fork(loadResource);
-    yield fork(saveResource);
-    yield fork(hideResource);
-    yield fork(addFieldToResource);
-    yield fork(changeFieldStatus);
-    yield fork(createResource);
-}
+export default [
```
Comment (fzaninotto) [inline src/app/js/public/resource/sagas/index.js]: please revert

**Review by fzaninotto — CHANGES_REQUESTED** [review]
To be rebased when previous PR is merged

File: src/app/js/admin/preview/sagas.js (hunk 6 above)
Comment (ThieryMichel) [reply src/app/js/admin/preview/sagas.js:6]: done

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**